### ISSUE TITLE
Croppie is very slow on mobile phones when cropping large images

### ISSUE THREAD

**Issue by HRworks-Dom** [Foliotek/Croppie#178]
We reproduced this issue with several different phones:

You can use the upload demo croppie on [your official demo page](https://foliotek.github.io/Croppie/) to reproduce it. Newer smartphones take pictures with quite big file sizes. This is an example picture we took with one of our devices, which has a file size of 3.2 MB.

![img_20160729_141104749](https://cloud.githubusercontent.com/assets/20727081/17253467/379bee36-55b1-11e6-8670-f1a117909823.jpg)

Try downloading it on a smartphone, then uploading it from there to the upload demo example croppie. Now try to zoom and move the picture around. You will notice that croppie gets very slow, loading the image in little chunks.

![screenshot_2016-07-29-17-30-28](https://cloud.githubusercontent.com/assets/20727081/17253909/09f64bd2-55b3-11e6-8bba-e776eb2c3dfb.png)

![screenshot_2016-07-29-17-30-31](https://cloud.githubusercontent.com/assets/20727081/17253908/09ec8f16-55b3-11e6-92cd-a7f036e792f3.png)

We were using Android devices with the latest Chrome version.

**Comment by thedustinsmith** [comment]
@HRworks-Dom - Can you try implementing the fix described here to see if it improves any of your speed issues? 
https://github.com/Foliotek/Croppie/issues/167#issuecomment-236856735

**Comment by xVin** [comment]
Hi,
The fix from #167 sadly didn't really improve the speed for us.
Is there anything else we could try?

**Comment by edagenais** [comment]
I'm running into very similar issues. Using an iPhone 6 Plus, when selecting large images, performing a zoom operation - or even just waiting for the image to load - regularly causes the browser tab to crash in webkit browsers.

In Safari, I get this error:

> A problem occurred with this webpage so it was reloaded.

Couple of things I'm thinking could help prevent this issue:
- Immediately scale the image down to a max width/height ratio before it's rendered to display.
- When performing zoom operations, allow for "stepped" scaling. The event listener is getting hammered when a user chooses to zoom in/out, which I would guess most phones would have trouble with given images of large dimensions.

@thedustinsmith, I also tried the fix for #167 to no avail.

Any further thoughts on how we can optimize things?

**Comment by thedustinsmith** [comment]
I would suggest trying to limit the amount of operations being performed during the zoom.

One quick thing to try would be to set `enforceBoundary: false` when initializing your croppie.  That reduces a bit of logic.

Removing any more logic in the `onZoom` function would require more research.

I'm not comfortable with enforcing a max-width/height restriction inside of croppie.  I'm leaning towards that being something that is handled by the developer using the API.

**Comment by edagenais** [comment]
@thedustinsmith, I'm not suggesting your force a max. I was suggesting to have optional parameters that down-scale the image to user specified dimensions before the image is rendered to the display.

I tested this by stuffing an image in an in-memory-canvas, resizing and then throwing that image over to croppie and the performance seem to improve for images of large dimensions.

**Comment by thedustinsmith** [comment]
Great!  Would you mind sharing some code so other people with the same problem can use it?

**Comment by edagenais** [comment]
Crudely:

```
var reader = new FileReader();
var img = new Image();
reader.onload = function ( e ) {
    img.onload = function () {
        var canvas = document.createElement( 'canvas' );
        var ctx = canvas.getContext( "2d" );
        var width = img.width;
        var height = img.height;
        var MAX_WIDTH = 1000;
        var MAX_HEIGHT = 1000;

        // Resize maintaining aspect ratio
        if ( width > height ) {
            if ( width > MAX_WIDTH ) {
                height *= MAX_WIDTH / width;
                width = MAX_WIDTH;
            }
        } else {
            if ( height > MAX_HEIGHT ) {
                width *= MAX_HEIGHT / height;
                height = MAX_HEIGHT;
            }
        }
        canvas.width = width;
        canvas.height = height;

        // Copy the image to the canvas and resize it.
        ctx.drawImage( img, 0, 0, width, height );

        // Toss the new down-scaled image over to Croppie
        $output.croppie( 'bind', {
                url: canvas.toDataURL( 'image/png' )
            } )
            .then(
                function ( value ) {
                    $output.croppie( 'setZoom', 0 );
                },
                function ( reason ) {
                    console.error( 'Image binding failed. Reason: ', reason );
                }
            );
    }
    img.src = e.target.result;
}
```

I was thinking the max width/height could be a new optional parameter to the bind method that the developer can set.

This totally resolved my problem with performance and crashes. Images load super fast and there's no lag when zooming and/or moving.

**Comment by BernardDoherty** [comment]
I agree, a max width/height as a new optional parameter would be useful.
I was able to use a solution similar to @edagenais. The only issue is it does not retain the EXIF information. Images taking on the iPhone are now rotated 90 degrees. A nice feature of Croppie is that it takes care of handling the EXIF orientation.

**Comment by edagenais** [comment]
@BernardDoherty, I actually leveraged [another library](https://github.com/blueimp/JavaScript-Load-Image) to manage the EXIF and pre-sizing and then tossed the output to Croppie for all the user manipulations. Worked out really well.

**Comment by MentalGear** [comment]
Wow, great, could someone of you guys try to build this into a PR ?
